cimi-resources/src/com/sixsq/slipstream/ssclj/resources/external_object/utils.clj

```diff
@@ -234,3 +234,41 @@
               (catch Exception _
                 (log/warn (str "Could not access the metadata for S3 object " object))))]
     (if md5 (assoc eo :md5sum md5) eo)))
+
+
+;; Function separated to allow for mocking in unit tests.
+(defn set-acl-public-read
+  [s3client bucket object]
+  (.setObjectAcl s3client bucket object CannedAccessControlList/PublicRead))
+
+
+(defn try-set-public-read-object
+  [s3client bucket object]
+  (try-catch-aws-fn (set-acl-public-read s3client bucket object)))
+
+
+(defn set-public-read-object
+  "Returns the untouched resource. Side effect is only on S3 permissions"
+  [{:keys [objectStoreCred bucketName objectName] :as resource}]
+  (let [s3client (-> objectStoreCred
+                     (format-creds-for-s3-api)
+                     (get-s3-client))]
+    (try-set-public-read-object s3client bucketName objectName)
+    resource))
+
+
+(defn public-url
```

Comment:
I don't think that this function has the behavior that you're expecting based on its use in the `add-public-url` function.

 - If the `.getURL` throws a runtime exception, then this will propagate up to the user.  Although, perhaps this will never throw.
 - If the URL is null, then the original `.toString` implementation will throw an NPE.  I've changed this to `str`, but in this case it will return an empty string for `(str nil)`, making the check in `add-public-url` useless.
 - It looks like `.getURL` will return a URL in all cases, whether the object is public or not.  Perhaps a more generic name for the `public-url` function would be more appropriate.